src/CoreWCF.Primitives/src/CoreWCF/Channels/MessageEncoderCompressionHandler.cs

```diff
@@ -13,45 +17,30 @@ internal static class MessageEncoderCompressionHandler
+        internal static async ValueTask<ReadOnlySequence<byte>> DecompressBufferAsync(ReadOnlySequence<byte> buffer, CompressionFormat compressionFormat, long maxReceivedMessageSize)
         {
-            MemoryStream memoryStream = new MemoryStream(buffer.Array, buffer.Offset, buffer.Count);
-            int maxDecompressedSize = (int)Math.Min(maxReceivedMessageSize, int.MaxValue);
+            if (buffer.Length > maxReceivedMessageSize)
+            {
+                throw Fx.Exception.AsError(new QuotaExceededException(SR.Format(SRCommon.MaxReceivedMessageSizeExceeded, maxReceivedMessageSize)));
+            }
 
-            using (BufferManagerOutputStream bufferedOutStream = new BufferManagerOutputStream(SRCommon.MaxReceivedMessageSizeExceeded, 1024, maxDecompressedSize, bufferManager))
+            Pipe pipe = new();
+            foreach (ReadOnlyMemory<byte> memory in buffer)
             {
-                bufferedOutStream.Write(buffer.Array, 0, buffer.Offset);
+                await pipe.Writer.WriteAsync(memory);
+            }
 
-                byte[] tempBuffer = bufferManager.TakeBuffer(DecompressBlockSize);
-                try
-                {
-                    using (Stream ds = compressionFormat == CompressionFormat.GZip ?
-                        (Stream)new GZipStream(memoryStream, CompressionMode.Decompress) :
-                        (Stream)new DeflateStream(memoryStream, CompressionMode.Decompress))
-                    {
-                        while (true)
-                        {
-                            int bytesRead = ds.Read(tempBuffer, 0, DecompressBlockSize);
-                            if (bytesRead > 0)
-                            {
-                                bufferedOutStream.Write(tempBuffer, 0, bytesRead);
-                            }
-                            else
-                            {
-                                break;
-                            }
-                        }
-                    }
-                }
-                finally
-                {
-                    bufferManager.ReturnBuffer(tempBuffer);
-                }
+            await pipe.Writer.FlushAsync();
+            await pipe.Writer.CompleteAsync();
 
-                byte[] decompressedBytes = bufferedOutStream.ToArray(out int length);
-                bufferManager.ReturnBuffer(buffer.Array);
-                buffer = new ArraySegment<byte>(decompressedBytes, buffer.Offset, length - buffer.Offset);
-            }
+            using Stream gZipStream =
+                compressionFormat == CompressionFormat.GZip
+                    ? new GZipStream(pipe.Reader.AsStream(), CompressionMode.Decompress)
+                    : new DeflateStream(pipe.Reader.AsStream(), CompressionMode.Decompress);
+
+            using var ms = new MemoryStream();
+            await gZipStream.CopyToAsync(ms);
```

Comment:
BufferManagerOutputStream is better performing. MemoryStream initially creates a byte[] with a size which is the power of 2 immediately larger than the first write. Then with each write, if it needs more space, it allocates a new array at least double the size of the current array (or if that's not big enough, the next power of 2 size) and copies the data over. This results in a lot of allocations and a lot of copying. BufferManagerOutputStream creates an initial buffer size that you specify, then when that's full, creates an additional buffer twice the size and starts filling that up. It keeps on allocating new buffers with no copying, but it uses the BuffferManager to do the allocation so hopefully no GC needed. When it's time to convert it to an array, it know how big of a buffer it needs so allocates the correct size byte[] and copies its contents over. A lot less copying, and a better allocation pattern.

Reply:
updated to use BufferManagerOutputStream